### PR TITLE
Fixes #16007

### LINKED ISSUE #16007: Inline if for variable bool value makes the result a reference/pointer to the data
### Zig Version

0.11.0-dev.3395+1e7dcaa3a

### Steps to Reproduce and Observed Behavior

When we use an inline if, the result value is a pointer to the data.

So if we say 
```zig
var condition1 = true;
const ans1 = if (condition1) answers[0] else answers[1]
```
then `ans1` will be a reference to `answers[0]`, and if we change `answers[0]`, then value in `ans1` will also change.

This does not happen if `condition1` is declared as `const`.

I initially caught the bug in version 2939, with variable optional types, but that has been fixed since.

Reproducible example:
```zig
const std = @import("std");

const Block = struct {
    value: u8,
};

pub fn main() !void {
    var blocks: [16]Block = undefined;
    {
        blocks[0] = .{ .value = 0 };
        // using variable optionals and bools
        var variable_opt_index: ?usize = 0;
        var variable_bool: bool = true;
        const variable_opt_index_block = if (variable_opt_index) |i| blocks[i] else blocks[1];
        const variable_bool_index_block: Block = if (variable_bool) blocks[0] else blocks[1];
        // using constant optionals and bools
        const const_opt_index: ?usize = 0;
        const const_bool: bool = true;
        const const_opt_index_block = if (const_opt_index) |i| blocks[i] else blocks[1];
        const const_bool_index_block = if (const_bool) blocks[0] else blocks[1];
        const opt_value = blocks[0].value;
        // changing the underlying value
        blocks[0].value = 42;
        std.debug.print("opt_value = {d}\n", .{opt_value});
        std.debug.print("variable_opt_index_block.value = {d} // should be 0\nvariable_bool_index_block.value = {d} // should be 0\n", .{
            variable_opt_index_block.value, variable_bool_index_block.value,
        });
        std.debug.print("const_opt_index_block.value = {d}\nconst_bool_index_block.value = {d}\n", .{
            const_opt_index_block.value, const_bool_index_block.value,
        });
    }
}
```

Output:
```
opt_value = 0
variable_opt_index_block.value = 0 // should be 0
variable_bool_index_block.value = 42 // should be 0
const_opt_index_block.value = 0
const_bool_index_block.value = 0
```

### Expected Behavior

The value returned should not be a reference. So changing the underlying data should not change the value stored.

Expected output for the above code:
```
opt_value = 0
variable_opt_index_block.value = 0 // should be 0
variable_bool_index_block.value = 0 // should be 0
const_opt_index_block.value = 0
const_bool_index_block.value = 0
```

### PR DESCRIPTION
**Pull request by arbrk1**
Fixes #16007 (and maybe some other related issues).

The llvm-codegen load elision procedure had been broken for a load-break pair. As I understand, usually such a pair has one of two forms:

```
1. load shouldn't be elided -- %a is used somewhere later
%b = load(..., %a)
...
%c!= br(..., %b!)

2. load can be elided -- %a is not used anywhere later
%b = load(..., %a!)
...
%c!= br(..., %b!)
```

Unfortunately, `fn categorizeOperand` looked only at the operand of the break instruction, which seems to be always dead. This PR changes the operand looked at by `fn categorizeOperand` to the operand of the load instruction which does matter.

### CODE REVIEW

File: test/behavior.zig [hunk 1]
```diff
@@ -143,6 +143,7 @@ test {
     _ = @import("behavior/bugs/13785.zig");
     _ = @import("behavior/bugs/14854.zig");
     _ = @import("behavior/bugs/15778.zig");
+    _ = @import("behavior/bugs/16007.zig");
```
Comment (Vexu) [inline test/behavior.zig:146]: Please do not add new `bugs/XXX.zig` files, instead put the tests in one of the existing categories.

**Review by Vexu — CHANGES_REQUESTED** [review]

File: test/behavior.zig (hunk 1 above)
Comment (arbrk1) [reply test/behavior.zig]: I'm sorry. Tests moved to `if.zig`

File: test/behavior/if.zig [hunk 2]
```diff
@@ -167,3 +167,42 @@ test "if-@as-if chain" {
...
 }
+
+test "if value shouldn't be load-elided if used later (structs)" {
+    const Foo = struct { x: i32 };
+
+    var a = Foo{ .x = 1 };
+    var b = Foo{ .x = 1 };
+
+    var condition = false;
+    std.mem.doNotOptimizeAway({
```
Comment (Vexu) [inline test/behavior/if.zig:178]: Passing `void` to `doNotOptimizeAway` is a noop.

File: test/behavior/if.zig [hunk 3]
```diff
@@ -167,3 +167,36 @@ test "if-@as-if chain" {
...
+    return true;
+}
+
+test "if value shouldn't be load-elided if used later (structs)" {
+    const Foo = struct { x: i32 };
+
+    var a = Foo{ .x = 1 };
+    var b = Foo{ .x = 1 };
+
+    const c = if (@call(.never_inline, returnTrue, .{})) a else b;
```
Comment (Vexu) [inline test/behavior/if.zig:181]: `never_inline` is not able to stop LLVM optimizations so maybe it'd be better to stick to using a `var` like in the issue.

Comment (arbrk1) [reply test/behavior/if.zig:181]: But as I understand, LLVM optimizations do not matter in this issue: the zig codegen had been making a wrong decision about load-elision (and consequently even `-O ReleaseFast` made `c.x` equal to 2).

I've tried to make the test somewhat future-proof relative to the possible changes in the zig codegen (e.g. if it started to do some basic optimizations before producing IR, the test would no longer be relevant with a simple `var cond = false; cond = true;`)

**Pull request merged**